utils/s3/client.cc

```diff
@@ -854,6 +858,167 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
+                        auto front = buffer->share(0, part_remainder);
+                        buffer->trim_front(part_remainder);
+                        // return the remainder of the buffer back to the pipe, so the next
+                        // part can pick it up
+                        _part_body.reader.unread(std::move(*buffer));
+                        co_await out.write(front.get(), front.size());
+                        written = part_remainder;
+                    } else {
+                        written = buffer->size();
+                        co_await out.write(buffer->get(), buffer->size());
+                    }
+                }
+                co_await out.flush();
+            } catch (...) {
+                ex = std::current_exception();
+            }
+            co_await out.close();
+            if (ex) {
+                std::rethrow_exception(std::move(ex));
+            }
+        });
+        auto holder = _bg_flushes.hold();
+        return _client->make_request(std::move(req), [this, part_size, part_number, start = s3_clock::now()] (group_client& gc, const http::reply& reply, input_stream<char>&& in_) mutable -> future<> {
+            auto etag = reply.get_header("ETag");
+            s3l.trace("uploaded {} part data -> etag = {} (upload id {})", part_number, etag, _upload_id);
+            _part_etags[part_number] = std::move(etag);
+            gc.write_stats.update(part_size, s3_clock::now() - start);
+            return make_ready_future();
+        }).handle_exception([this, part_number] (auto ex) {
+            s3l.warn("couldn't upload part {}: {} (upload id {})", part_number, ex, _upload_id);
+        }).finally([holder = std::move(holder)] {});
+    }
+
+    static size_t calc_part_size(size_t total_size, size_t maximum_part_size) {
+        assert(maximum_part_size >= aws_minimum_part_size);
+        for (size_t part_size = maximum_part_size; part_size >= aws_minimum_part_size; part_size -= 1_MiB) {
+            size_t num_parts = (total_size + part_size) / part_size;
+            if (num_parts <= aws_maximum_parts_in_piece) {
+                return part_size;
+            }
+        }
+        s3l.error("unable to figure out part_size: total_size={}, maximum_part_size={}",
+                  total_size, maximum_part_size);
+        throw std::domain_error(fmt::format("total_size={}, maximum_part_size={}",
+                                            total_size, maximum_part_size));
+    }
+
+    future<> multipart_upload() {
+        // a background job which consumes the buffers pushed to part_body,
+        // and write them to with multipart upload
+        co_await start_upload();
+        size_t part_size = 0;
+        for (size_t offset = 0; offset < _total_size; offset += part_size) {
+            part_size = std::min(_total_size - offset, _part_size);
+            co_await upload_part(offset, part_size);
```

Comment:
The parts upload parallelism is lost here. But it's related to [this](https://github.com/scylladb/scylladb/pull/16387/commits/e204025d1d3d9a31fc008b446403ed00839b2eab#r1642980014) observation

Reply:
yeah. i admit that with current design, we don't have parallelism when writing to a certain ranged sink.